ast/ast.go

```diff
@@ -12,19 +12,72 @@ import (
 	"grol.io/grol/token"
 )
 
+type Priority int8
+
+const (
+	_ Priority = iota
+	LOWEST
+	ASSIGN      // =
+	OR          // ||
+	AND         // &&
+	EQUALS      // ==
+	LESSGREATER // > or <
+	SUM         // +
+	PRODUCT     // *
+	DIVIDE      // /
+	PREFIX      // -X or !X
+	CALL        // myFunction(X)
+	INDEX       // array[index]
+	DOTINDEX    // map.str access
+)
+
+var Precedences = map[token.Type]Priority{
+	token.ASSIGN:     ASSIGN,
+	token.OR:         OR,
+	token.AND:        AND,
+	token.COLON:      AND, // range operator and maps (lower than lambda)
+	token.EQ:         EQUALS,
+	token.NOTEQ:      EQUALS,
+	token.LAMBDA:     EQUALS,
+	token.LT:         LESSGREATER,
+	token.GT:         LESSGREATER,
+	token.LTEQ:       LESSGREATER,
+	token.GTEQ:       LESSGREATER,
+	token.PLUS:       SUM,
+	token.MINUS:      SUM,
+	token.BITOR:      SUM,
+	token.BITXOR:     SUM,
+	token.BITAND:     PRODUCT,
+	token.ASTERISK:   PRODUCT,
+	token.PERCENT:    PRODUCT,
+	token.LEFTSHIFT:  PRODUCT,
+	token.RIGHTSHIFT: PRODUCT,
+	token.SLASH:      DIVIDE,
+	token.INCR:       PREFIX,
+	token.DECR:       PREFIX,
+	token.LPAREN:     CALL,
+	token.LBRACKET:   INDEX,
+	token.DOT:        DOTINDEX,
+}
+
+//go:generate stringer -type=Priority
+var _ = DOTINDEX.String() // force compile error if go generate is missing.
+
 type PrintState struct {
-	Out             io.Writer
-	IndentLevel     int
-	ExpressionLevel int
-	IndentationDone bool // already put N number of tabs, reset on each new line
-	Compact         bool // don't indent at all (compact mode), no newlines, fewer spaces, no comments
-	prev            Node
-	last            string
+	Out                  io.Writer
+	IndentLevel          int
+	ExpressionPrecedence Priority
+	IndentationDone      bool // already put N number of tabs, reset on each new line
+	Compact              bool // don't indent at all (compact mode), no newlines, fewer spaces, no comments
+	AllParens            bool // print all expressions fully parenthesized.
+	prev                 Node
+	last                 string
 }
 
 func DebugString(n Node) string {
 	ps := NewPrintState()
 	ps.Compact = true
+	ps.AllParens = true
```

Comment:
I should have raised it earlier but using Parens as an abbreviatipn for Parenthesis is not obvious

Reply:
<img width="526" alt="Screenshot 2024-08-27 at 7 54 16 AM" src="https://github.com/user-attachments/assets/6efd6d15-985a-4a36-aecd-07442ccfd062">

Reply:
Ok.

![4eb4e00f8f9fb00cb5b4e80b86ff49ba045b8022bc0aa13a76f5343140ab4c52.jpg](https://github.com/user-attachments/assets/42431011-4401-4be6-888a-813ac509d358)

Reply:
lol :)